### ISSUE TITLE
Loading performance? how to order the files loading to client?

### ISSUE THREAD

**Issue by PolGuixe** [gatsbyjs/gatsby#2049]
I am building my first Gatsby site and I am experiencing some slow loading in production. The CSS seems slow to load. 

The site: http://same-writer.surge.sh/
The repo is here: https://github.com/Zetoff/portfolio/tree/feature/project-pages

Is there a way to load the  CSS first? or show a loading component in the mean time?

**Comment by markmichon** [comment]
Aside from the images (which could be optimized, as they are the real bottleneck issue), it looks like the weird flash of unstyled content is limited to your navbar component. Perhaps it's something with material-ui components? The rest of the components and their styles are loading immediately (within 400ms at least in chrome's network profiler)

**Comment by PolGuixe** [comment]
I'll try to investigate why the Appbar's CSS takes so long to generate. 

Yeap, images need to be optimized ;)

**Comment by cr101** [comment]
@PolGuixe You might need to use `sheetsRegistry` as per the [Material-UI documentation](https://material-ui-1dab0.firebaseapp.com/guides/server-rendering/)

> We must provide to the page the needed style. It's important that we provide the page with the required CSS, otherwise the page will render with just the HTML then wait for the CSS to be injected by the client, causing it to flicker.

**Comment by PolGuixe** [comment]
I'll try to implement it today! I'll posts the results.

**Comment by cr101** [comment]
@PolGuixe Please do. I'm also planning to use Material-UI v1.0 on a new project. 
I noticed that Material-UI v1.0.0-**beta.8** is out so you might want to upgrade as well.

**Comment by PolGuixe** [comment]
@cr101 I've used Material-UI pre v.1 in few projects. This is the first time I am using the new v.1. the API has changed a lot and performance-wise seems a lot better. 

I'll upgrade to latest beta as well.

**Comment by cr101** [comment]
There is also [starter examples for CRA and NextJS](https://material-ui-1dab0.firebaseapp.com/getting-started/examples/) which might help. 
It would be great to also have a Material-UI v1 example project for Gatsby.

**Comment by PolGuixe** [comment]
@cr101 are you familiar with Gatsby's SSR API? 

I need to use `replaceRenderer`, I don't understand when this is executed? on every page? 

I am setting the `MuiThemeProvider` in the layout component. 

I am going to try to export the `replaceRenderer` from there.

**Comment by PolGuixe** [comment]
I have implemented it.... not sure whether it works or not... I need more understanding of Gatsby's API or more testing 😅 

@cr101 would you mind having a look https://github.com/Zetoff/portfolio/pull/6

It is hosted here: http://abounding-zephyr.surge.sh/

**Comment by PolGuixe** [comment]
Definitely, it is not working. The functions are not executed by Gatsby.

**Comment by cr101** [comment]
@PolGuixe I had a look but can't figure it out sorry.

Has anyone else used Material-UI v1-beta with Gatsby?

**Comment by PolGuixe** [comment]
@wzrdzl I see you did the JSS plugin. Can you throw some light on this issue? Thanks 😄

**Comment by sebastienfi** [comment]
Here is your optimized image.
[optimizilla.zip](https://github.com/gatsbyjs/gatsby/files/1289997/optimizilla.zip)

Do you guys want to PR an example site with Material-UI ?

**Comment by PolGuixe** [comment]
@sebastienfi I want it to work properly. If it does then we could add it as an example. 😉

**Comment by PolGuixe** [comment]
@KyleAMathews could you help us?

How should I access the `replaceRenderer` and the `onInitialClientRender` API?

**Comment by KyleAMathews** [comment]
* https://www.gatsbyjs.org/docs/ssr-apis/#replaceRenderer
* https://www.gatsbyjs.org/docs/browser-apis/#onInitialClientRender

**Comment by PolGuixe** [comment]
@KyleAMathews  Where should I export the functions from from? 

e.g.
- `project/src/replaceRenderer.js`
- `project/src/onInitialClientRender.js`

can it be exported from `project/src/layouts/index.js`?

**Comment by markmichon** [comment]
As far as I understand it, SSR APIs only run from gatsby-ssr.js, and browser only run in gatsby-browser, etc.

**Comment by tinrab** [comment]
Any updates on this? I'm having the same problem with FOUC on https://outcrawl.com/. It's using [material-ui](https://github.com/callemall/material-ui).

**Comment by PolGuixe** [comment]
I'll try to have a look at it this week.

**Comment by mhyrr** [comment]
@tinrab What was your solution?

I'm having a similar issue when using Styled Components.  It looks like the Components built using this library load later.  Example http://justadream.co/

**Comment by tinrab** [comment]
@mhyrr This works for material-ui: use [gatsby-plugin-jss](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-plugin-jss) plugin, then set `sheetsManager` to an empty map on `MuiThemeProvider`, like [here](https://github.com/outcrawl/site/blob/master/src/layouts/index.jsx#L22). Or what @sogasg suggested.

**Comment by migs540** [comment]
@mhyrr I have the same problem as I see on your site. 

What we see on the screen:

homepage loads (with a css keyframe transition) then after a second or so the css keyframe transition is repeated (it shouldn't repeat) resulting in a kind of flash like on your site. 

What Developer tools, network shows:

The first keyframe transition (remembering we should only see this once) appears when layouts-index, pages-index, path-index, app and common js scripts run (and the html for the index page is loaded). 

Then we wait a second or so (slowed down using 3G slow emulation) and see the rest of the gatsby site loading and immediately the keyframe is triggered again:

path---otherpages
component---otherpages
etc.

These have an initiator of bootstrap and for some reason at the same time the keyframe associated with index.js is triggered to make the animation appear to users for the second time.  It's a clunky user experience and I imagine it goes away if the keyframe is removed? 

Any ideas on how to get keyframes running just the once? I don't think there's anything wrong with the css, for instance: I gzipped all my js files (meaning they didn't run) and with just html it works (just the once as intended). 

React is designed to re-render just parts of the screen / update just parts of the dom - any idea why these keyframes are being triggered twice?

**Comment by jonathan-chin** [comment]
@sogasg this fix was exactly what I was looking for. just had to add one line to gatsby-config and it started working.

**Comment by oliviertassinari** [comment]
If that can help, Material-UI host a [fully functional example](https://github.com/mui-org/material-ui/tree/1669a30422329f2a5467008792e515db8eb8da55/examples/gatsby) with Gatsby.

**Comment by oliviertassinari** [comment]
@KyleAMathews I think that we can close the issue.

**Comment by mpolinowski** [comment]
@oliviertassinari Thank you for the example for Material-UI!

It works for me in development but running into issues with build.

I cloned the repo, run npm install (all @latest ) and gatsby build.

```
error Building static HTML for pages failed
```

```
10 | 'use strict';var m=require("object-assign"),n=require("fbjs/lib/emptyObject"),p=require("fbjs/lib/emptyFunction"),q="function"===typeof Symbol&&Symbol["
for"],r=q?Symbol["for"]("react.element"):60103,t=q?Symbol["for"]("react.call"):60104,u=q?Symbol["for"]("react.return"):60105,v=q?Symbol["for"]("react.portal"):
60106,w=q?Symbol["for"]("react.fragment"):60107,x="function"===typeof Symbol&&Symbol.iterator;
> 11 | function y(a){for(var b=arguments.length-1,e="Minified React error #"+a+"; visit http://facebook.github.io/react/docs/error-decoder.html?invariant\x3d"+
a,c=0;c<b;c++)e+="\x26args[]\x3d"+encodeURIComponent(arguments[c+1]);b=Error(e+" for the full message or use the non-minified dev environment for full errors a
nd additional helpful warnings.");b.name="Invariant Violation";b.framesToPop=1;throw b;}
     |
                                                                  ^
```

```
WebpackError: Minified React error #143; visit http://facebook.github.io/react/docs/error-decoder.html?invariant=143 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
```

**Comment by oliviertassinari** [comment]
@mpolinowski I can't reproduce the issue.
> success Building CSS — 6.079 s
success Building production JavaScript bundles — 13.326 s
success Building static HTML for pages — 4.396 s

**Comment by mpolinowski** [comment]
@oliviertassinari 

I got your repository to work by first installing Gatsby and then adding react/react-dom @latest. But this is overwriting Gatsby's react/react-dom v15.6 dependency and seems to cause a lot of issues on my actual project.

I tried to just install Gatsby (with react/react-dom v15.6) and use your solution with it - but no luck.

Also, I noticed that there is an official [gatsby-plugin-jss](https://www.gatsbyjs.org/packages/gatsby-plugin-jss/) that could substitute the jss and react-jss dependency.

I will have to look more into that issue.
